### PR TITLE
gh #43 Python test development

### PR DESCRIPTION
**Pull request by Anbukannadhasan**
Python tests for tvSettings L3 test cases

### CODE REVIEW

File: host/tests/tvSettings_L3_Tests/tvSettings_test28_GammaTable.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                # Generate Red, Green, and Blue values based on size
+                size = combination["size"]
+                red = combination["red"]
+                green = combination["green"]
+                blue = combination["blue"]
+
+                self.log.stepStart(f'Setting Gamma Table Size: {size}, R: {red}, G: {green}, B: {blue}')
+
+                # Set the Gamma Table
+                self.testtvSettings.setGammaTable(size, red, green, blue)
```
Comment (bhanucbp) [inline host/tests/tvSettings_L3_Tests/tvSettings_test28_GammaTable.py:124]: setGammaTable takes only single value, but here list is sent to the function

File: host/tests/tvSettings_L3_Tests/tvSettings_test30_ComponentSaturation.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+        # Initialize the tvSettings module
+        self.testtvSettings.initialise()
+
+        # Get available TV Data Colors
+        tvDataColors = self.testtvSettings.getTVDataColor()
+
+        for stream in self.testStreams:
+            # Start the stream playback
+            self.testPlayer.play(stream)
```
Comment (bhanucbp) [inline host/tests/tvSettings_L3_Tests/tvSettings_test30_ComponentSaturation.py:81]: is the play duration enough for testing all saturation values? If not move play inside for loop

File: host/tests/tvSettings_L3_Tests/tvSettings_test32_ComponentLuma.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+dir_path = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(dir_path, "../"))
+
+from tvSettings_L3_Tests.tvSettingsHelperClass import tvSettingsHelperClass
+
+class tvSettings_test32_ComponentLuma(tvSettingsHelperClass):
+
+    # Predefined luma values ranging from 1 to 30
+    lumaValues = [1, 5, 10, 15, 30]
+    rackDevice = "dut"
```
Comment (bhanucbp) [inline host/tests/tvSettings_L3_Tests/tvSettings_test32_ComponentLuma.py:35]: remove this, this variable is already part of helper function

File: host/tests/tvSettings_L3_Tests/tvSettings_test34_SetGammaPattern.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                    redLevels = self.ten_bit_levels
+                    greenLevels = self.ten_bit_levels
+                    blueLevels = self.ten_bit_levels
+                else:  # 8-bit depth case
+                    redLevels = self.eight_bit_levels
+                    greenLevels = self.eight_bit_levels
+                    blueLevels = self.eight_bit_levels
+
+                # Iterate over the predefined levels in order
+                for redLevel, greenLevel, blueLevel in zip(redLevels, greenLevels, blueLevels):
```
Comment (bhanucbp) [inline host/tests/tvSettings_L3_Tests/tvSettings_test34_SetGammaPattern.py:126]: loop is confusing, try to make it simple

File: host/tests/tvSettings_L3_Tests/tvSettings_test40_EnableDynamicContrast.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+        for stream in self.testStreams:
+            # Start the stream playback
+            self.testPlayer.play(stream)
+
+            for contrast in self.dynamicContrastLevels:
+                self.log.stepStart(f'Query Dynamic Contrast Level: {contrast} Stream: {stream}')
+
+                # Query the user before setting the contrast level
+                if self.testQueryDynamicContrastLevel(contrast):
```
Comment (bhanucbp) [inline host/tests/tvSettings_L3_Tests/tvSettings_test40_EnableDynamicContrast.py:100]: why user query required here?

**Review by bhanucbp — COMMENTED** [review]
address review comments

File: host/tests/Classes/tvSettings.py [hunk 6]
```diff
@@ -0,0 +1,2180 @@
...
+    def __init__(self, deviceProfilePath:str, session=None, devicePath="/tmp" ):
+        """
+        Initializes the tvSettings class function.
+        """
+        self.deviceProfile = ConfigRead( deviceProfilePath, self.moduleName)
+        self.suitConfig    = ConfigRead(self.menuConfig, self.moduleName)
+        self.utMenu        = UTSuiteNavigatorClass(self.menuConfig, self.moduleName, session, devicePath)
+        self.testSession   = session
+        self.utMenu.start()
+        self.formatChangeCB = self.suitConfig.get("test").get("callback").get("formatChange_status")
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:227]: removed the formatChange_staus. As I am not using it anymore

File: host/tests/Classes/tvSettings.py [hunk 7]
```diff
@@ -0,0 +1,2180 @@
...
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Initialize")
+
+    def checkVideoFormat(self):
+        """
+        Gets the current video format.
+
+        Returns:
+            None
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:255]: Done

File: host/tests/Classes/tvSettings.py [hunk 8]
```diff
@@ -0,0 +1,2180 @@
...
+            },
+            {
+                "query_type": "direct",
+                "query": f"Enter {gammaTableSize} gamma values for Blue (0 - 1023):",
+                "input": " ".join(map(str, gammaBlue))
+            }
+        ]
+
+        # Display prompts and collect user inputs
+        result = self.utMenu.select(self.testSuite, "Save Gamma Table", promptWithAnswers)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:956]: Updated

File: host/tests/Classes/tvSettings.py [hunk 9]
```diff
@@ -0,0 +1,2180 @@
...
+        Returns:
+            None
+        """
+        promptWithAnswers = [
+            {
+                "query_type": "direct",
+                "query": "Enter Gray level (0 - 255):",
+                "input": str(grayPattern)
+            }
+    ]
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:1204]: Done

File: host/tests/Classes/tvSettings.py [hunk 10]
```diff
@@ -0,0 +1,2180 @@
...
+        Args:
+            None.
+
+        Returns:
+            list: List of supported TV Data Colors.
+        """
+
+        outTVDataColors = []
+        for color in tvDataColor:
+                outTVDataColors.append(tvDataColor(color).name)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:2066]: Done

File: host/tests/Classes/tvSettings.py [hunk 11]
```diff
@@ -0,0 +1,2180 @@
...
+
+        Returns:
+            None: If the function fails to retrieve the callback status.
+            tuple: The height, width, interlaced status, resolution, and user data, as indicated by the system callback.
+        """
+        result = self.testSession.read_until("Received Video Resolution Change callback Frame heightxwidth")
+
+        formatCallbackPattern = r"Received Video Resolution Change callback Frame heightxwidth:\[(\d+)x(\d+)\], IsInterlaced\[(\d+)\] Res:\[(\w+)\], userData\[\]\[0x([0-9A-Fa-f]+)\]"
+
+        return self.searchPattern(result, formatCallbackPattern)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:2150]: tested and modified the pattern

File: host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+sys.path.append(os.path.join(dir_path, "../../"))
+
+from raft.framework.plugins.ut_raft.configRead import ConfigRead
+from raft.framework.plugins.ut_raft.utPlayer import utPlayer
+from raft.framework.plugins.ut_raft.utUserResponse import utUserResponse
+from raft.framework.plugins.ut_raft import utHelperClass
+from raft.framework.core.logModule import logModule
+from Classes.tvSettings import tvSettingsClass  # Changed from tvSettingClass
+
+class tvSettingsHelperClass(utHelperClass):  # Changed from tvSettingHelperClass
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py:37]: Done

File: host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+            log (class, optional): Parent log class. Defaults to None.
+        """
+        super().__init__(testName, qcId, log)
+
+        # Load test setup configuration
+        self.testSetup = ConfigRead(self.testSetupPath, self.moduleName)
+
+        # Open Sessions for player, secondary player and for hal test
+        self.player_session = self.dut.getConsoleSession("ssh_player")
+        self.secondary_player_session = self.dut.getConsoleSession("ssh_player_secondary")
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py:59]: Not required and removed it

File: host/tests/Classes/tvSettings_test_suite.yml [hunk 14]
```diff
@@ -3,7 +3,13 @@
...
-        execute: "" #Execute command
+        artifacts:
+          - "../../../bin/hal_test"
+          - "../../../bin/libut_control.so"
+          - "../../../bin/run.sh"
+        execute:
+          command: "run.sh"
+          arguments: ""
         type: UT-C # C (UT-C Cunit) / C++ (UT-G (g++ ut-core gtest backend))
         suites:
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings_test_suite.yml:14]: updated L2. L1 is containing different Banks. Need Discussion on this

File: host/tests/tvSettings_L3_Tests/tvSettings_test02_CheckVideoResolution.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+            streamFullPath = os.path.join(self.deviceDownloadPath, os.path.basename(streamUrl))
+
+            # Start the stream playback
+            self.testPlayer.play(streamFullPath)
+            time.sleep(3)
+
+            self.log.stepStart(f'Video Resolution Level: {resolution} Stream: {streamFullPath}')
+
+            self.log.stepStart(f'Video Resolution {resolution} Callback Test')
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test02_CheckVideoResolution.py:89]: updated

File: host/tests/tvSettings_L3_Tests/tvSettings_test06_BacklightFade.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+        for stream in self.testStreams:
+            for fromVal in self.backlightValues:
+                for toVal in self.backlightValues:
+                    for fadeDuration in self.fadeDurations:
+                        self.log.stepStart(f'BacklightFade from {fromVal} to {toVal} with duration {fadeDuration} ms Stream:{stream}')
+
+                        # Start the stream playback before setting backlight fade
+                        self.testPlayer.play(stream)
+                        time.sleep(3)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test06_BacklightFade.py:96]: Discussed

File: host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+dir_path = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(dir_path, "../"))
+
+from tvSettings_L3_Tests.tvSettingsHelperClass import tvSettingsHelperClass
+
+class tvSettings_test21_ColorTempRgain(tvSettingsHelperClass):
+
+    rgainValues = [0, 512, 1024, 1536, 2047]
+    saveSetFlags = [0, 1]  # Save only or set only flags
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py:28]: Removed

File: host/tests/Classes/tvSettings.py [hunk 18]
```diff
@@ -0,0 +1,2180 @@
...
+        Gets the TV Dimming Modes.
+
+        Args:
+            None.
+
+        Returns:
+            bool : Dimming Modes
+        """
+        # Get the Back light Mode index from the profile
+        TVDimmingModes = self.deviceProfile.get("DimmingMode").get("index")
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:1808]: Done

File: host/tests/tvSettings_L3_Tests/tvSettings_test18_DynamicGamma.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+import time
+
+dir_path = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(dir_path, "../"))
+
+from tvSettings_L3_Tests.tvSettingsHelperClass import tvSettingsHelperClass
+
+class tvSettings_test18_DynamicGamma(tvSettingsHelperClass):
+
+    dynamicGammaLevels = [1.80, 1.90, 2.00, 2.10, 2.20, 2.30, 2.40, 2.50, 2.60]
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test18_DynamicGamma.py:34]: Done

File: host/tests/tvSettings_L3_Tests/tvSettings_test19_DolbyVision.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+        for stream in self.testStreams:
+            # Start the stream playback
+            self.testPlayer.play(stream)
+
+            for DolbyVision in self.testtvSettings.getDolbyVisionInfo():
+                self.log.stepStart(f'Dolby Vision Level: {DolbyVision} Stream: {stream}')
+
+                # Query if Dolby Vision level should be applied
+                result = self.testQueryDolbyVisionLevel(DolbyVision, True)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test19_DolbyVision.py:107]: Removed

File: host/tests/tvSettings_L3_Tests/tvSettings_test20_PictureMode.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+
+        for stream in self.testStreams:
+            # Start the stream playback
+            self.testPlayer.play(stream)
+
+            for pictureMode in self.testtvSettings.getPictureModeInfo():
+                self.log.stepStart(f'Picture Mode Level: {pictureMode} Stream: {stream}')
+
+                # Query if Picture Mode level should be applied
+                result = self.testQueryPictureModeLevel(pictureMode, True)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test20_PictureMode.py:104]: Removed

File: host/tests/tvSettings_L3_Tests/tvSettings_test10_Brightness.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+
+                result = self.testVerifyBrightnessLevel(brightness, True)
+
+                self.log.stepResult(result, f'Brightness Level: {brightness} Stream: {stream}')
+
+            # Stop the stream playback
+            self.testPlayer.stop()
+
+        # Terminate tvSettings Module
+        self.testtvSettings.terminate()
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test10_Brightness.py:103]: Done

File: host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+
+            for colorTemperature in colorTemperatureValues:
+                for rgain in self.rgainValues:
+                    self.log.stepStart(f'Save Operation - Color Temperature: {colorTemperature} Rgain: {rgain} Stream: {stream}')
+
+                    # Set the Color Temperature and Rgain (Save flag = 1 means Save operation)
+                    self.testtvSettings.setRgainValue(colorTemperature, rgain, 1)
+
+                    # Stop the current stream
+                    self.testPlayer.stop()
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py:102]: modified

File: host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+
+            # Get the Color Temperature values
+            colorTemperatureValues = self.testtvSettings.getColorTemperatureInfo()
+
+            for colorTemperature in colorTemperatureValues:
+                for rgain in self.rgainValues:
+                    self.log.stepStart(f'Save Operation - Color Temperature: {colorTemperature} Rgain: {rgain} Stream: {stream}')
+
+                    # Set the Color Temperature and Rgain (Save flag = 1 means Save operation)
+                    self.testtvSettings.setRgainValue(colorTemperature, rgain, 1)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py:99]: As discussed modified

File: host/tests/tvSettings_L3_Tests/tvSettings_test30_ComponentSaturation.py (hunk 2 above)
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test30_ComponentSaturation.py:81]: Will make the stream length as per the test duration

File: host/tests/tvSettings_L3_Tests/tvSettings_test32_ComponentLuma.py (hunk 3 above)
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test32_ComponentLuma.py:35]: removed

File: host/tests/tvSettings_L3_Tests/tvSettings_test34_SetGammaPattern.py (hunk 4 above)
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test34_SetGammaPattern.py:126]: corrected. Please check

File: host/tests/tvSettings_L3_Tests/tvSettings_test40_EnableDynamicContrast.py (hunk 5 above)
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettings_test40_EnableDynamicContrast.py:100]: Removed

File: host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (bhanucbp) [reply host/tests/tvSettings_L3_Tests/tvSettings_test21_ColorTempRgain.py:28]: not removed

**Comment by bhanucbp** [comment]
remove hal_test from bin/obj folder

**Review by bhanucbp — COMMENTED** [review]
address review comments

File: host/install.sh [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+# Setup raft
+clone_repo git@github.com:rdkcentral/python_raft.git "${RAFT_DIR}" "1.0.0" "in ./raft"
+install_pip_requirements "${RAFT_DIR}"/requirements.txt
+
+# Setup rdk-raft
+#clone_repo git@github.com:rdkcentral/rdk_raft.git "${PLUGINS_FRAMEWORK_DIR}/rdk_raft" develop "in ./raft/framework/plugins"
+#install_pip_requirements "${TESTS_DIR}"/rdk_raft/requirements.txt
+
+# Setup ut-raft
+clone_repo git@github.com:rdkcentral/ut-raft.git "${PLUGINS_FRAMEWORK_DIR}/ut_raft" develop "in ./raft/framework/plugins"
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/install.sh:169]: Done

File: host/install.sh [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+## Setup and start venv
+setup_and_enable_venv
+
+### Clone required repos ###
+# Setup raft
+clone_repo git@github.com:rdkcentral/python_raft.git "${RAFT_DIR}" "1.0.0" "in ./raft"
+install_pip_requirements "${RAFT_DIR}"/requirements.txt
+
+# Setup rdk-raft
+#clone_repo git@github.com:rdkcentral/rdk_raft.git "${PLUGINS_FRAMEWORK_DIR}/rdk_raft" develop "in ./raft/framework/plugins"
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/install.sh:165]: Done

File: host/tests/Classes/tvSettings.py [hunk 27]
```diff
@@ -0,0 +1,2302 @@
...
+    VIDEO_SOURCE_RESERVED                 = auto()          # Video source reserved
+    VIDEO_SOURCE_RESERVED1                = auto()          # Video source reserved
+    VIDEO_SOURCE_RESERVED2                = auto()          # Video source reserved
+    VIDEO_SOURCE_RESERVED3                = auto()          # Video source reserved
+    VIDEO_SOURCE_MAX                      = auto()          # End of enum
+
+
+class tvSettingsClass():
+
+    moduleName = "tvSettings"
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:209]: Done

File: host/tests/Classes/tvSettings.py [hunk 28]
```diff
@@ -0,0 +1,2302 @@
...
+
+    This module provides common extensions for tvSettings Module.
+    """
+    def __init__(self, deviceProfilePath:str, session=None, devicePath="/tmp" ):
+        """
+        Initializes the tvSettings class function.
+        """
+        self.deviceProfile = ConfigRead( deviceProfilePath, self.moduleName)
+        self.suitConfig    = ConfigRead(self.menuConfig, self.moduleName)
+        self.utMenu        = UTSuiteNavigatorClass(self.menuConfig, self.moduleName, session, devicePath)
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py]: Done

File: host/tests/Classes/tvSettings.py [hunk 29]
```diff
@@ -0,0 +1,2302 @@
...
+        Enables the tvSettings port.
+
+        Args:
+            None.
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select(self.testSuite, "Terminate")
+
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings.py:2302]: Done

File: host/tests/Classes/tvSettings_test_suite.yml [hunk 30]
```diff
@@ -3,7 +3,13 @@
 tvSettings:  # Prefix must always exist
     description: "tvSettings testing profile / menu system for UT"
     test:
-        execute: "" #Execute command
+        artifacts:
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/Classes/tvSettings_test_suite.yml:6]: Done

File: host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py [hunk 31]
```diff
@@ -0,0 +1,212 @@
...
+
+        # Get path to device profile file
+        self.deviceProfile = os.path.join(dir_path, self.cpe.get("test").get("profile"))
+
+        self.deviceDownloadPath = self.cpe.get("target_directory")
+
+    def testDownloadAllAssets(self):
+        """
+        Downloads the test artifacts and streams listed in the test setup configuration.
+        This function retrieves TV settings configurations and other necessary files and
```
[the remark this answers is not in the document]
Comment (Anbukannadhasan) [reply host/tests/tvSettings_L3_Tests/tvSettingsHelperClass.py:77]: Done

File: host/tests/tvSettings_L3_Tests/tvSettings_test02_CheckVideoResolution.py [hunk 32]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+#** *****************************************************************************
+# *
+# * License and copyright notice as above
```
Comment (mhughesacn) [inline host/tests/tvSettings_L3_Tests/tvSettings_test02_CheckVideoResolution.py:4]: Please add a proper header here.

**Review by bhanucbp — APPROVED** [review]
ship it

**Review by KarthikeyanR470 — APPROVED** [review]
Approved

**Pull request merged**